### PR TITLE
Adx 640 restricted delays

### PR DESCRIPTION
**Pull request by jonathansberry**
The hide_inaccessible_resources flag is still causing significant delays in package_search requests. 

I figured that in these cases there are two actions that are repeated for every resource when in practise they only need to be done once per request or package. 

- checking the user can access package_show only needs to be done once per package, and in actual fact this is done when fetching the package dict already, higher up the code stack.
- Checking the users org list is repeated for every resource, this only needs to be done once for the request.

With these two changes package_search requests are reduced to 25% of the time that they previously took. 

I'm concerned about a software engineer coming along later and calling `_restricted_resource_list_accessible_by_user` with the package_dict argument populated without having authed to check the user can see package_show. (e.g.t hey may have fetched the package_dict straight from the model). However adding in a package show check adds 0.12 seconds per package in the search result (5 s over the existing 50 packages in CKAN).

### CODE REVIEW

File: ckanext/restricted/logic.py [hunk 1]
```diff
@@ -98,16 +100,17 @@ def restricted_check_user_resource_access(user, resource_dict, package_dict, use
         return {'success': True}
 
     # Get organization list
-    user_organization_dict = {}
+    user_organization_dict = getattr(user_obj, 'org_dict', {})
```
Comment (tomeksabala) [inline ckanext/restricted/logic.py:103]: I think it's possible that here `user_obj` in `None`

**Review by tomeksabala — CHANGES_REQUESTED** [review]

File: ckanext/restricted/logic.py (hunk 1 above)
Comment (jonathansberry) [reply ckanext/restricted/logic.py:103]: That's fine isn't it? That's in part the purpose of getattr right? i.e...
```
>>> getattr(None, 'org_dict', {})
{}
```

File: ckanext/restricted/logic.py [hunk 2]
```diff
@@ -68,10 +68,12 @@ def restricted_get_restricted_dict(resource_dict):
     return restricted_dict
 
 
-def restricted_check_user_resource_access(user, resource_dict, package_dict, user_obj=None):
+def restricted_check_user_resource_access(user, resource_dict, package_dict, user_obj=None, pkg_show=True):
```
Comment (tomeksabala) [inline ckanext/restricted/logic.py:71]: `check_access_package_show`

File: ckanext/restricted/action.py [hunk 3]
```diff
@@ -114,16 +114,37 @@ def _restricted_resource_list_accessible_by_user(context, resource_list, package
     restricted_resources_list = []
     user_name = logic.restricted_get_username_from_context(context)
     user_obj = context.get('auth_user_obj')
+    if user_obj:
+        user_obj.org_dict = _get_org_dict(user_name)
```
Comment (tomeksabala) [inline ckanext/restricted/action.py:118]: convert to pass `user_org_dict` as `logic.restricted_check_user_resource_access` optional parameter

File: ckanext/restricted/action.py [hunk 4]
```diff
@@ -114,16 +114,37 @@ def _restricted_resource_list_accessible_by_user(context, resource_list, package
...
+            user_obj=user_obj,
+            pkg_show=False
+        )
+        if auth_result.get('success', False):
             restricted_resources_list.append(resource_dict)
 
     return restricted_resources_list
 
 
+def _get_org_dict(user_name):
```
Comment (tomeksabala) [inline ckanext/restricted/action.py:136]: Let's make it public function in the `logic` module in order to remove code duplication.

File: ckanext/restricted/action.py [hunk 5]
```diff
@@ -114,16 +114,37 @@ def _restricted_resource_list_accessible_by_user(context, resource_list, package
...
             package_dict = package_show(context, {'id': resource_dict['package_id']})
-        if logic.restricted_check_user_resource_access(user_name, resource_dict, package_dict, user_obj=user_obj).get('success', False):
+        auth_result = logic.restricted_check_user_resource_access(
+            user_name,
+            resource_dict,
+            package_dict,
+            user_obj=user_obj,
+            pkg_show=False
+        )
+        if auth_result.get('success', False):
```
Comment (tomeksabala) [inline ckanext/restricted/action.py:130]:
```
        user_has_resource_access = logic.restricted_check_user_resource_access(
            user_name,
            resource_dict,
            package_dict,
            user_obj=user_obj,
            pkg_show=False
        ).get('success', False)
        if user_has_resource_access:
```

**Review by tomeksabala — CHANGES_REQUESTED** [review]

**Review by tomeksabala — APPROVED** [review]

**Pull request merged**